### PR TITLE
feat!: add support for selecting the Dokka type to be included in the publication

### PR DESCRIPTION
**Pull request by gciatto**
- remove useless class `MavenRepositoryDescriptor` (class `Repository` should be sufficient)

- allow users to select which Dokka task to use when generating `javadoc` artifacts to be published (dafault is still `dokkaHtml`, so nothing changes by default)

In case you were wondering why the second feature is needed:
```
:full:dokkaJavadoc (Thread[Execution worker Thread 4,5,main]) started.
Exception in thread "Thread-217" org.jetbrains.dokka.DokkaException: Pre-generation validity check failed: Dokka Javadoc plugin currently does not support generating documentation for multiplatform project. Please, adjust your configuration
	at org.jetbrains.dokka.base.generation.SingleModuleGeneration.validityCheck(SingleModuleGeneration.kt:93)
	at org.jetbrains.dokka.base.generation.SingleModuleGeneration.generate(SingleModuleGeneration.kt:25)
	at org.jetbrains.dokka.DokkaGenerator$generate$1.invoke(DokkaGenerator.kt:29)
	at org.jetbrains.dokka.DokkaGenerator$generate$1.invoke(DokkaGenerator.kt:17)
	at org.jetbrains.dokka.DokkaGeneratorKt.timed(DokkaGenerator.kt:76)
	at org.jetbrains.dokka.DokkaGeneratorKt.access$timed(DokkaGenerator.kt:1)
	at org.jetbrains.dokka.DokkaGenerator.generate(DokkaGenerator.kt:22)
	at org.jetbrains.dokka.DokkaBootstrapImpl.generate(DokkaBootstrapImpl.kt:58)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jetbrains.dokka.gradle.DelegatedInvocationHandler.invoke(automagicTypedProxy.kt:40)
	at com.sun.proxy.$Proxy128.generate(Unknown Source)
	at org.jetbrains.dokka.gradle.AbstractDokkaTask$generateDocumentation$1$1.run(AbstractDokkaTask.kt:208)
```

### CODE REVIEW

File: src/main/kotlin/org/danilopianini/gradle/mavencentral/Configuration.kt [hunk 1]
```diff
@@ -75,7 +75,7 @@ fun Project.configureRepository(repoToConfigure: Repository) {
         publishing.repositories { repository ->
             repository.maven { mavenArtifactRepository ->
                 mavenArtifactRepository.name = repoToConfigure.name
-                mavenArtifactRepository.url = URI(repoToConfigure.url)
+                mavenArtifactRepository.url = repoToConfigure.url.map { URI(it) }.get()
```
Comment (DanySK) [inline src/main/kotlin/org/danilopianini/gradle/mavencentral/Configuration.kt:78]: Why this change? It looks like you are breaking the laziness...

File: src/main/kotlin/org/danilopianini/gradle/mavencentral/PublishOnCentralExtension.kt [hunk 2]
```diff
@@ -87,24 +88,32 @@ open class PublishOnCentralExtension(val project: Project) {
         repoOwner.map { "https://github.com/$it/${project.name}" }
     )
 
+    /**
+     * The style of `javadoc` artifacts being published on Maven repositories.
+     */
+    val docStyle: Property<DocStyle> = project.objects.property<DocStyle>().convention(
+        DocStyle.HTML
```
Comment (DanySK) [inline src/main/kotlin/org/danilopianini/gradle/mavencentral/PublishOnCentralExtension.kt:95]:
```suggestion
        DocStyle.JAVADOC
```
Am I correct?

File: src/main/kotlin/org/danilopianini/gradle/mavencentral/PublishOnCentralExtension.kt [hunk 3]
```diff
@@ -87,24 +88,32 @@ open class PublishOnCentralExtension(val project: Project) {
...
+    )
+
+    /**
+     * Selects all Dokka tasks aimed at generating documentation compliant with the style declared by [docStyle].
+     */
+    internal val dokkaTasks: TaskCollection<Task>
+        get() = project.tasks.matching { task ->
+            task.name.let { it.startsWith("dokka") && it.endsWith(docStyle.get().name, ignoreCase = true) }
+        }
```
Comment (DanySK) [inline src/main/kotlin/org/danilopianini/gradle/mavencentral/PublishOnCentralExtension.kt:104]: Fetching tasks by name is very fragile. If someone defines a random task called "dokkaFoo" it would be returned -- I think it might be better to use the task type

**Review by DanySK — CHANGES_REQUESTED** [review]
Good job. There are a few points that I need explanation for.

File: src/main/kotlin/org/danilopianini/gradle/mavencentral/PublishOnCentralExtension.kt (hunk 2 above)
Comment (gciatto) [reply src/main/kotlin/org/danilopianini/gradle/mavencentral/PublishOnCentralExtension.kt:95]: Oh yes. I was thinking to my default, not yours. Apologies. JAVADOC is the correct default

File: src/main/kotlin/org/danilopianini/gradle/mavencentral/Configuration.kt (hunk 1 above)
Comment (gciatto) [reply src/main/kotlin/org/danilopianini/gradle/mavencentral/Configuration.kt:78]: I see no real alternative here: we're assigning a property to an ordinary mutable variable, so we need to access the final actual of the property.

Btw my edit was about merging classes `MavenRepositoryDescriptor` and `Repository` which were essentially the same thing, yet having no common super type. Here, the previous code was using the `url` variable of the former class, which was not supporting laziness. I replaced this variable by an homonymous property from the latter class, yet the logic of the code is unchanged: so, the property will certainly have a value when reaching this point in the code.

However, if you have a better way to write this, please enlighten me. Is the a way to add a callback to a property being assigned?

File: src/main/kotlin/org/danilopianini/gradle/mavencentral/PublishOnCentralExtension.kt (hunk 3 above)
Comment (gciatto) [reply src/main/kotlin/org/danilopianini/gradle/mavencentral/PublishOnCentralExtension.kt]: Fixed in b8bf993

**Comment by gciatto** [comment]
@DanySK please approve the workflow

**Comment by gciatto** [comment]
@nicolasfara could you please comment the feature I introduced in ec6709e ?

Facts: 
- currently, Dokka does not support the generation of Javadoc-styled sites for multi-platoform projects
- Maven Central specification requires published artifacts to include a `javadoc` jar including Javadoc-compliant specification

IMHO, the workaround should be: make it possible for multi-plaform project to use Dokka's Html documentation instead. 
I know this is a violation of MC's specification, but to me it seems better than having no documentation at all.

Notice that Sonatype's automatic checks simply ensure that a file named `<artifact>-javadoc-<version>.jar` is uploaded. They do not look into the jar, meaning that uploading a jar containing Html documentation is technically acceptable

File: src/main/kotlin/org/danilopianini/gradle/mavencentral/PublishOnCentralExtension.kt (hunk 2 above)
Comment (gciatto) [reply src/main/kotlin/org/danilopianini/gradle/mavencentral/PublishOnCentralExtension.kt]: I force pushed a new version of the commit where the convention is set to JAVADOC

**Comment by nicolasfara** [comment]
> @nicolasfara could you please comment the feature I introduced in [3a2e1f4](https://github.com/DanySK/publish-on-central/commit/3a2e1f48e35d395545c713e5bf14eee991cc0056)?
> 
> Facts:
> 
> * currently, Dokka does not support the generation of Javadoc-styled sites for multi-platoform projects
> * Maven Central specification requires published artifacts to include a `javadoc` jar including Javadoc-compliant specification
> 
> IMHO, the workaround should be: make it possible for multi-plaform project to use Dokka's Html documentation instead. I know this is a violation of MC's specification, but to me it seems better than having no documentation at all.
> 
> Notice that Sonatype's automatic checks simply ensure that a file named `<artifact>-javadoc-<version>.jar` is uploaded. They do not look into the jar, meaning that uploading a jar containing Html documentation is technically acceptable.
> 
> For all these reasons, my edit is aimed at letting the plugin user choose which particular documentation style should be included in Maven publications. Of course, the convention would be to use the Javadoc style by default.

From my side should be ok!

**Comment by DanySK** [comment]
OK. I'm convinced. Please document it in the readme and I'll rebase this in.

**Comment by gciatto** [comment]
@DanySK done
